src/lib/layout/layout.ts

```diff
@@ -80,6 +80,50 @@ type OutputInfo = {
   startModuleInPath: string
   tsOutputDir: string
   bundleOutputDir: string | null
+  /**
+   * The final path to the start module. When bundling disbaled, same as `startModuleOutPath`.
+   */
+  startModule: string
+  /**
+   * The final output dir. If bundler is enabled then this is `bundleOutputDir`.
+   * Otherwise it is `tsOutputDir`.
+   *
+   * When bundle case, this accounts for the bundle environment, which makes it
+   * **DIFFERENT** than the source root. For example:
+   *
+   * ```
+   * <out_root>/node_modules/
+   * <out_root>/api/app.ts
+   * <out_root>/api/index.ts
+   * ```
+   */
+  root: string
+  /**
+   * If bundler is enabled then the final output dir where the **source** is
+   * located. Otherwise same as `tsOutputDir`.
+   *
+   * When bundle case, this is different than `root` because it tells you where
+   * the source starts, not the build environment.
+   *
+   * For example, here `source_root` is `<out_root>/api` becuase the user has
+   * set their root dir to `api`:
+   *
+   * ```
+   * <out_root>/node_modules/
+   * <out_root>/api/app.ts
+   * <out_root>/api/index.ts
+   * ```
+   *
+   * But here, `source_root` is `<out_root>` because the user has set their root
+   * dir to `.`:
+   *
+   * ```
+   * <out_source_root>/node_modules/
+   * <out_source_root>/app.ts
+   * <out_source_root>/index.ts
+   * ```
+   */
+  sourceRoot: string
```

Comment:
This feels a bit hairy 🙈

Reply:
Yeah... The goal here is to move toward making as much as the codebase agnostic to if the build has a bundle step or not. Eventually all of it. I know you already thought about this. What I was trying to avoid was maintaining repeated if conditions in many many places, not to mention path joining the derivations and so on.

I think we need to keep the refactoring pressure on for sure though.